src/socket.rs

```diff
@@ -0,0 +1,358 @@
+    /// single filter matching all incoming frames is installed.
+    pub fn set_filters(&self, filters: &[CanFilter]) -> io::Result<()> {
+        set_socket_option_mult(self.fd, SOL_CAN_RAW, CAN_RAW_FILTER, filters)
+    }
+
+    /// Sets the error mask on the socket.
+    ///
+    /// By default (`ERR_MASK_NONE`) no error conditions are reported as
+    /// special error frames by the socket. Enabling error conditions by
+    /// setting `ERR_MASK_ALL` or another non-empty error mask causes the
+    /// socket to receive notification about the specified conditions.
+    #[inline]
+    pub fn set_error_mask(&self, mask: u32) -> io::Result<()> {
+        set_socket_option(self.fd, SOL_CAN_RAW, CAN_RAW_ERR_FILTER, &mask)
+    }
+
+    /// Enable or disable loopback.
+    ///
+    /// By default, loopback is enabled, causing other applications that open
+    /// the same CAN bus to see frames emitted by different applications on
+    /// the same system.
+    #[inline]
+    pub fn set_loopback(&self, enabled: bool) -> io::Result<()> {
+        let loopback: c_int = if enabled { 1 } else { 0 };
+        set_socket_option(self.fd, SOL_CAN_RAW, CAN_RAW_LOOPBACK, &loopback)
+    }
+
+    /// Enable or disable receiving of own frames.
+    ///
+    /// When loopback is enabled, this settings controls if CAN frames sent
+    /// are received back immediately by sender. Default is off.
+    pub fn set_recv_own_msgs(&self, enabled: bool) -> io::Result<()> {
+        let recv_own_msgs: c_int = if enabled { 1 } else { 0 };
+        set_socket_option(self.fd, SOL_CAN_RAW, CAN_RAW_RECV_OWN_MSGS, &recv_own_msgs)
+    }
+
+    /// Enable or disable join filters.
+    ///
+    /// By default a frame is accepted if it matches any of the filters set
+    /// with `set_filters`. If join filters is enabled, a frame has to match
+    /// _all_ filters to be accepted.
+    pub fn set_join_filters(&self, enabled: bool) -> io::Result<()> {
+        let join_filters: c_int = if enabled { 1 } else { 0 };
+        set_socket_option(self.fd, SOL_CAN_RAW, CAN_RAW_JOIN_FILTERS, &join_filters)
+    }
+}
+
+impl AsRawFd for CanSocket {
+    fn as_raw_fd(&self) -> RawFd {
+        self.fd
+    }
+}
+
+impl FromRawFd for CanSocket {
+    unsafe fn from_raw_fd(fd: RawFd) -> CanSocket {
+        CanSocket { fd, }
+    }
+}
+
+impl IntoRawFd for CanSocket {
+    fn into_raw_fd(self) -> RawFd {
+        self.fd
+    }
+}
+
+impl Drop for CanSocket {
+    fn drop(&mut self) {
+        self.close().ok(); // ignore result
+    }
+}
+
+
+/// CanFilter
+///
+/// Contains an internal id and mask. Packets are considered to be matched by
+/// a filter if `received_id & mask == filter_id & mask` holds true.
+#[derive(Debug, Copy, Clone)]
+#[repr(C)]
+pub struct CanFilter {
+    _id: u32,
+    _mask: u32,
+}
+
+impl CanFilter {
+    /// Construct a new CAN filter.
+    pub fn new(id: u32, mask: u32) -> Result<CanFilter, ConstructionError> {
```

Comment:
Shouldn't this now be defined with the `StandardId` and `ExtendedId` types to be consistent? Or would that make things unnecessarily complicated?

Previously, I was thinking to add a `From<(u32,u32)>` (from a u32 tuple) for this to make it easier to create a list of filters. Maybe do both?